File: C++/depth-first_search/depth-first_search.cpp
//
// Created by hunor on 31.07.2023.
//

#include "depth-first_search.h"

DFSGraph::DFSGraph(int number_of_nodes) : number_of_nodes(number_of_nodes) {
    adjacency_list.resize(number_of_nodes + 1); // Adjust for 1-based indexing
}

void DFSGraph::add_edge_directed(int first_node, int second_node) {
    adjacency_list[first_node].push_back(second_node);
}

void DFSGraph::add_edge_nondirected(int first_node, int second_node) {
    adjacency_list[first_node].push_back(second_node);
    adjacency_list[second_node].push_back(first_node);
}

void DFSGraph::depth_first_search(int start_node) {
    vector<bool> visited(number_of_nodes + 1, false); // Track visited nodes
    dfs_helper(start_node, visited);
    cout << endl;
}

void DFSGraph::dfs_helper(int current_node, vector<bool> &visited) {
    visited[current_node] = true;
    cout << current_node << " ";
    for (auto neighbor: adjacency_list[current_node]) {
        if (!visited[neighbor]) {
            dfs_helper(neighbor, visited);
        }
    }
}

File: TypeScript/src/depth-first-search/depth-first-search.ts
import { Queue } from 'queue-typescript'; // Import a queue library for TypeScript

/**
 * Represents a class for performing Depth-First Search (DFS) on a graph.
 */
export class DFSGraph {
  private numberOfNodes: number; // The number of nodes in the graph
  private adjacencyList: number[][]; // Adjacency list representing the graph

  /**
   * Constructor for the DFSGraph class.
   * @param numberOfNodes The number of nodes in the graph.
   */
  constructor(numberOfNodes: number) {
    this.numberOfNodes = numberOfNodes;
    this.adjacencyList = new Array<number[]>(numberOfNodes + 1).fill([]);
  }

  /**
   * Add an undirected edge between two nodes in the graph.
   * @param firstNode The first node.
   * @param secondNode The second node.
   */
  addEdgeNonDirected(firstNode: number, secondNode: number): void {
    this.adjacencyList[firstNode].push(secondNode);
    this.adjacencyList[secondNode].push(firstNode);
  }

  /**
   * Add a directed edge from the first node to the second node in the graph.
   * @param firstNode The source node.
   * @param secondNode The target node.
   */
  addEdgeDirected(firstNode: number, secondNode: number): void {
    this.adjacencyList[firstNode].push(secondNode);
  }

  /**
   * Perform Depth-First Search (DFS) starting from a specified node.
   * @param startNode The starting node for DFS traversal.
   */
  depthFirstSearch(startNode: number): void {
    const visited: boolean[] = new Array<boolean>(this.numberOfNodes + 1).fill(false);
    this.dfsHelper(startNode, visited);
    console.log();
  }

  private dfsHelper(currentNode: number, visited: boolean[]): void {
    visited[currentNode] = true;
    process.stdout.write(`${currentNode} `);

    for (const neighbor of this.adjacencyList[currentNode]) {
      if (!visited[neighbor]) {
        this.dfsHelper(neighbor, visited);
      }
    }
  }
}
